server/src/main/java/org/opensearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -835,6 +865,89 @@ public void clusterStatePublished(ClusterChangedEvent clusterChangedEvent) {
+                    MetadataCreateIndexService.updateRemoteStoreSettings(
+                        indexSettingsBuilder,
+                        remoteStoreRepoNames.v1(),
+                        remoteStoreRepoNames.v2()
+                    );
+
+                    // Overriding all existing customs with only path and path hash algorithm based settings
+                    Map<String, String> newCustomRemoteStoreMetadata = new HashMap<>();
+                    if (remoteStorePathStrategyResolver != null) {
+                        RemoteStorePathStrategy newPathStrategy = remoteStorePathStrategyResolver.get();
+                        newCustomRemoteStoreMetadata.put(RemoteStoreEnums.PathType.NAME, newPathStrategy.getType().name());
+                        if (Objects.nonNull(newPathStrategy.getHashAlgorithm())) {
+                            newCustomRemoteStoreMetadata.put(
+                                RemoteStoreEnums.PathHashAlgorithm.NAME,
+                                newPathStrategy.getHashAlgorithm().name()
+                            );
+                        }
+                    }
+
+                    final IndexMetadata.Builder builder = IndexMetadata.builder(indexMetadata)
+                        .settings(indexSettingsBuilder)
+                        .putCustom(REMOTE_STORE_CUSTOM_KEY, newCustomRemoteStoreMetadata);
+                    builder.settingsVersion(1 + indexMetadata.getVersion());
+                    metadataBuilder.put(builder);
+                }
+            }
+            return ClusterState.builder(state).metadata(metadataBuilder).build();
+        }
+
+        /**
+         * Returns <code>true</code> iff all shards for the index are in `STARTED` state
+         * all those `STARTED` shard copies are present in remote store enabled nodes
+         *
+         * @param indexRoutingTable current {@link IndexRoutingTable} from cluster state
+         * @param discoveryNodes set of discovery nodes from cluster state
+         * @param indexMetadata current {@link IndexMetadata} from cluster state
+         * @return <code>true</code> or <code>false</code> depending on the met conditions
+         */
+        private boolean needsRemoteIndexSettingsUpdate(
+            IndexRoutingTable indexRoutingTable,
+            DiscoveryNodes discoveryNodes,
+            IndexMetadata indexMetadata
+        ) {
+            assert indexMetadata != null : "IndexMetadata for a STARTED shard cannot be null";
+            boolean allStartedShardsOnRemote = indexRoutingTable.allShardsStarted()
+                && (indexRoutingTable.shardsMatchingPredicateCount(ShardRouting::started) == indexRoutingTable.shardsMatchingPredicateCount(
+                    shardRouting -> discoveryNodes.get(shardRouting.currentNodeId()).isRemoteStoreNode()
+                ));
+            return allStartedShardsOnRemote && IndexMetadata.INDEX_REMOTE_STORE_ENABLED_SETTING.get(indexMetadata.getSettings()) == false;
```

Comment:
The check of `(indexRoutingTable.shardsMatchingPredicateCount(ShardRouting::started) == indexRoutingTable.shardsMatchingPredicateCount(
                    shardRouting -> discoveryNodes.get(shardRouting.currentNodeId()).isRemoteStoreNode()` is not accurate , as it is not account for started shards for checking shards on remote store . We should check just for all started shards that they are on remote node.

Reply:
We also need to make sure that all the primary shards are assigned as well . There can be case when 1 shard with 0 replica is unassigned , but that copy is present on a docrep node, which is temporarily down.